p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/FtpTransportProtocolHandler.java

```diff
@@ -0,0 +1,200 @@
+package org.eclipse.tycho.p2maven.transport;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.net.ftp.*;
+import org.codehaus.plexus.component.annotations.Component;
+import org.codehaus.plexus.component.annotations.Requirement;
+import org.codehaus.plexus.logging.Logger;
+import org.codehaus.plexus.personality.plexus.lifecycle.phase.Disposable;
+import org.eclipse.tycho.MavenRepositorySettings.Credentials;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Handles files discovery over the FTP protocol.
+ *
+ * @author Edoardo Luppi
+ */
+@Component(role = TransportProtocolHandler.class, hint = "ftp")
+public class FtpTransportProtocolHandler implements TransportProtocolHandler, Disposable {
+    private final Map<String, FTPClient> clients = new ConcurrentHashMap<>(8);
+
+    @Requirement
+    private Logger logger;
+
+    @Requirement
+    private TransportCacheConfig cacheConfig;
+
+    @Requirement
+    private MavenAuthenticator authenticator;
+
+    @Override
+    public long getLastModified(final URI uri) throws IOException {
+        final FTPClient client = getClient(uri);
+
+        if (!client.hasFeature(FTPCmd.MDTM)) {
+            logger.debug("Could not retrieve the last modification timestamp for: " + uri);
+            return -1;
+        }
+
+        final FTPFile file = client.mdtmFile(uri.getPath());
+
+        if (file == null) {
+            throw new IOException("Could not find file: " + uri);
```

Comment:
This should be FileNotFoundException then

Reply:
Changed.